File: Comp Coding Practice/DMOJ/CCC '20 S2 - Escape Room.cpp
#include <bits/stdc++.h>

using namespace std;

// Initializing the vector and arrays
int M, N;
vector<vector<int>> room(1000001);
bool visited[1000001];

void dfs(int x)
{
    if (visited[x])
    {
        return;
    }
    // Checking if we have reached exit or not
    if (x == N * M)
    {
        cout << "yes" << endl;
        exit(0);
    }
    // Setting the cell as visited
    visited[x] = 1;
    for (int &t : room[x])
    {
        dfs(t);
    }
}

int main()
{

    cin >> M >> N;
    for (int i = 1; i <= M; i++)
    {
        for (int j = 1; j <= N; j++)
        {
            int x;
            cin >> x;
            room[i * j].push_back(x);
        }
    }
    dfs(1);
    cout << "no" << endl;
}

File: Comp Coding Practice/DMOJ/CCC '20 S2 - Escape Room.py


# Breadth First Search
# Starting from end point and checking if you can reach the entry point
# ! TLE for last couple batches

import sys
import collections

rows, cols = int(sys.stdin.readline()), int(sys.stdin.readline())
grid = []
for _ in range(rows):
    grid.append(list(map(int, sys.stdin.readline().split())))


def bfs(r, c):
    q = collections.deque()
    # setting grid value to "V" if it has been visited
    grid[r][c] = "V"
    q.append((r, c))
    while q:
        row, col = q.popleft()
        calc = (row + 1)*(col + 1)

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] != "V" and calc == grid[r][c]:
                    print(grid)

                    # This means we reached the starting point
                    if r == 0 and c == 0:
                        return "yes"
                    q.append((r, c))
                    grid[r][c] = "V"
    return "no"


lastRow, lastCol = rows - 1, cols - 1
print(bfs(lastRow, lastCol))
